### ISSUE TITLE
Incorrect interaction between unions containing nil and optional class fields

### ISSUE THREAD

**Issue by justarandomgeek** [EmmyLuaLs/emmylua-analyzer-rust#1195]
```lua
---@class Thing1
---@field thing1 true

---@class Thing2
---@field thing2 true

---@alias MaybeThing1 Thing1|nil

---@class ComplexThing1
---@field maybe? MaybeThing1

---@alias SomeThing ComplexThing1|Thing2

---@type SomeThing
local _c = {}
```

Hovers on `ComplexThing1` show the field `maybe` has been made required, which causes the `local _c` to incorrectly warn that it cannot accept the empty table, which is a perfectly valid `ComplexThing1`, and thus a member of `SomeThing`
